src/txn/txn.c

```diff
@@ -826,45 +827,52 @@ __wt_txn_commit(WT_SESSION_IMPL *session, const char *cfg[])
 		 * For prepared transactions commit timestamp could be earlier
 		 * than stable timestamp.
 		 */
-		WT_ERR(__wt_txn_commit_timestamp_validate(
-		    session, "commit", ts, &cval, !prepare));
 		txn->commit_timestamp = ts;
+		WT_ERR(__wt_txn_commit_timestamp_validate(session));
 		__wt_txn_set_commit_timestamp(session);
-		if (!prepare)
-			txn->durable_timestamp = txn->commit_timestamp;
+		txn->durable_timestamp = txn->commit_timestamp;
 	}
 
 	if (prepare && !F_ISSET(txn, WT_TXN_HAS_TS_COMMIT))
 		WT_ERR_MSG(session, EINVAL,
 		    "commit_timestamp is required for a prepared transaction");
 
 	/*
-	 * Durable timestamp is required for a prepared transaction.
-	 * If durable timestamp is not given, commit timestamp will be
-	 * considered as durable timestamp. We don't flag error if durable
-	 * timestamp is not specified for prepared transactions, but will flag
-	 * error if durable timestamp is specified for non-prepared
+	 * Durable timestamp is required only for a prepared transaction.
+	 * Return error if durable timestamp is not specified for a prepared
+	 * transaction and if durable timestamp is specified for non-prepared
 	 * transactions.
 	 */
 	WT_ERR(__wt_config_gets_def(
 	    session, cfg, "durable_timestamp", 0, &cval));
 	if (cval.len != 0) {
-		if (!prepare)
-			WT_ERR_MSG(session, EINVAL,
-			    "durable_timestamp should not be given for "
-			    "non-prepared transaction");
-
 		WT_ERR(__wt_txn_parse_timestamp(
 		    session, "durable", &ts, &cval));
 		/* Durable timestamp should be later than stable timestamp. */
-		F_SET(txn, WT_TXN_HAS_TS_DURABLE);
 		txn->durable_timestamp = ts;
-		WT_ERR(__wt_txn_commit_timestamp_validate(
-		    session, "durable", ts, &cval, true));
+		WT_ERR(__wt_txn_durable_timestamp_validate(session));
+		F_SET(txn, WT_TXN_HAS_TS_DURABLE);
+	}
+
+	if (prepare) {
+		if (!F_ISSET(txn, WT_TXN_HAS_TS_DURABLE))
```

Comment:
I believe this could be folded into the `if` above, if you make a change to the following else as well - and I think that would be an improvement.